File: sdk/java/src/main/java/com/aligntech/sdk/AlignTechClient.java
package com.aligntech.sdk;

import com.fasterxml.jackson.databind.ObjectMapper;

public class AlignTechClient implements AutoCloseable {

    private final ConfigStore store;
    private final Evaluator evaluator;
    private final ConfigPoller poller;

    private AlignTechClient(AlignTechConfig config) {
        this.store = new ConfigStore();
        this.evaluator = new Evaluator();
        ObjectMapper mapper = new ObjectMapper();
        this.poller = new ConfigPoller(config.baseUrl, config.appKey,
                config.pollInterval, store, mapper);

        poller.fetchInitial();
        poller.start();
    }

    /**
     * Returns true if the given flag evaluates to "enabled" for this user.
     * If the flag is unknown, returns false (safe default).
     */
    public boolean isEnabled(String flagKey, AlignTechUser user) {
        FlagConfig flag = store.get(flagKey);
        if (flag == null) return false;
        return evaluator.evaluate(flag, user).enabled();
    }

    /**
     * Returns the evaluation result for a flag, or null if the flag is unknown.
     */
    public Evaluator.EvaluationResult evaluate(String flagKey, AlignTechUser user) {
        FlagConfig flag = store.get(flagKey);
        if (flag == null) return null;
        return evaluator.evaluate(flag, user);
    }

    public boolean isReady() {
        return !store.isEmpty();
    }

    /**
     * Propagate a W3C traceparent header to the server on snapshot requests.
     * Call this with the current trace context from your application's tracer
     * so the server can correlate SDK calls with your request traces.
     *
     * @param traceParent W3C traceparent value (e.g. "00-{traceId}-{spanId}-01")
     */
    public void setTraceParent(String traceParent) {
        poller.setTraceParent(traceParent);
    }

    @Override
    public void close() {
        poller.close();
    }

    public static Builder builder() { return new Builder(); }

    public static class Builder {
        private AlignTechConfig config;

        public Builder config(AlignTechConfig c) { this.config = c; return this; }

        public AlignTechClient build() {
            if (config == null) throw new IllegalArgumentException("config is required");
            return new AlignTechClient(config);
        }
    }
}

File: sdk/dotnet/src/AlignTech.FeatureSDK/AlignTechClient.cs
using AlignTech.FeatureSDK.Models;

namespace AlignTech.FeatureSDK;

public class AlignTechClient : IDisposable
{
    private readonly ConfigStore _store;
    private readonly Evaluator _evaluator;
    private readonly ConfigPoller _poller;

    public AlignTechClient(AlignTechConfig config)
    {
        if (string.IsNullOrEmpty(config.AppKey))
        {
            throw new ArgumentException("AppKey is required", nameof(config));
        }

        _store = new ConfigStore();
        _evaluator = new Evaluator();
        _poller = new ConfigPoller(config, _store);
    }

    public async Task InitializeAsync()
    {
        await _poller.FetchInitial();
        _poller.Start();
    }

    public bool IsEnabled(string flagKey, AlignTechUser user)
    {
        var flag = _store.Get(flagKey);
        if (flag == null) return false;

        return _evaluator.Evaluate(flag, user).Enabled;
    }

    public EvaluationResult? Evaluate(string flagKey, AlignTechUser user)
    {
        var flag = _store.Get(flagKey);
        if (flag == null) return null;

        return _evaluator.Evaluate(flag, user);
    }

    public bool IsReady()
    {
        return !_store.IsEmpty();
    }

    public void SetTraceParent(string traceParent)
    {
        _poller.SetTraceParent(traceParent);
    }

    public void Close()
    {
        _poller.Stop();
    }

    public void Dispose()
    {
        _poller.Dispose();
        GC.SuppressFinalize(this);
    }
}
